Add conversions between DataType and ScalarType.

ezyang opened the pull request:
Differential Revision: D9298048

gchanan commented on aten/src/ATen/core/ScalarType.h:
```diff
@@ -38,6 +39,27 @@ enum class ScalarType {
+
+  switch(scalar_type) {
+    AT_FORALL_SCALAR_TYPES(DEFINE_CASE)
+    default: return DataType::uninitialized();
```
what does uninitialized represent?  The asymmetry is a bit surprising here, I would guess both functions either throw or return some sentinel type.

gchanan replied:
that sounds fine, just write that explicitly.

gchanan commented on aten/src/ATen/core/ScalarType.h:
```diff
@@ -38,6 +39,28 @@ enum class ScalarType {
+  if (dtype == caffe2::TypeMeta::Id<ctype>()) { \
+    return ScalarType:: name; \
+  }
+  AT_FORALL_SCALAR_TYPES(DEFINE_IF)
```
won't this still be asymmetric if someone passes in DataType::unitialized()?  (i.e. you won't return ScalarType::Undefined).

ezyang committed:
Add DataType alias

Differential Revision: D9346040

fbshipit-source-id: 5d25f118c1926fd242ad0229fc5f4fdd17724dca

ezyang committed:
Move all operator<< overloads out of the global namespace.

Summary:
Have you ever written an operator<< overload in the caffe2 namespace
in a core Caffe2 header, and then been stunned when some completely
unrelated code started breaking?  This diff fixes this problem!

The problem looks like this:
1. You're building against a really old version of glog (think 0.3.2,
   or something like that)
2. This version of glog defines operator<< overloads for std containers
   in the global namespace
3. You add a new overload in your current namespace (e.g., caffe2).
   Congratulations: this overload is *preferentially* chosen over
   the global namespace one for all calls to << in that namespace.
   And since it doesn't actually have std::vector overloads, unrelated
   Caffe2 code breaks.

Newer versions of glog have a fix for this: they have the line:

  namespace std { using ::operator<<; }

in their header.  So let's help old versions of glog out and do this ourselves.

In our new world order, operator<< overloads defined in the global namespace
won't work (unless they're for std containers, which work because of ADL).
So this diff also moves all those overloads to the correct namespace.

Differential Revision: D9344540

fbshipit-source-id: 258f7eadfa1969267ef198a60c1886a761d094d3

ezyang committed:
Add conversions between DataType and ScalarType.

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/10472

Reviewed By: gchanan

Differential Revision: D9298048

fbshipit-source-id: 79bcf6107fa32af442a9d1263ea0c8a2c80be637
